### ISSUE TITLE
wgsl: consider disabling out of bounds memory accesses to shared memories

### ISSUE THREAD

**Issue by dneto0** [gpuweb/gpuweb#3893]
Currently the spec says (at [Invalid Memory Reference](https://gpuweb.github.io/gpuweb/wgsl/#invalid-mem-reference)):

> [Stores](https://gpuweb.github.io/gpuweb/wgsl/#statement-assignment) to an invalid reference may do one of:
when the [originating variable](https://gpuweb.github.io/gpuweb/wgsl/#originating-variable) is a [uniform buffer](https://gpuweb.github.io/gpuweb/wgsl/#uniform-buffer) or a [storage buffer](https://gpuweb.github.io/gpuweb/wgsl/#storage-buffer), store the value to any [memory location(s)](https://gpuweb.github.io/gpuweb/wgsl/#memory-locations) of the WebGPU [GPUBuffer](https://gpuweb.github.io/gpuweb/#gpubuffer) bound to the [originating variable](https://gpuweb.github.io/gpuweb/wgsl/#originating-variable)

> when the [originating variable](https://gpuweb.github.io/gpuweb/wgsl/#originating-variable) is not a [uniform buffer](https://gpuweb.github.io/gpuweb/wgsl/#uniform-buffer) or [storage buffer](https://gpuweb.github.io/gpuweb/wgsl/#storage-buffer), store the value to any [memory locations(s)](https://gpuweb.github.io/gpuweb/wgsl/#memory-locations) in the originating variable

> not be executed

The first clause is intended to allow use of hardware-based robustness guarantees. (e.g. Vulkan [robustBufferAccess](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VkPhysicalDeviceFeatures))

The second clause is intended to allow stores to use clamped indices.

However, performing stores while clamping indices can generate data races.

For example:
```
var<workgroup> a: array<i32,10>;

@compute @workgroup_size(11)
fn main(@builtin(local_invocation_index) id: u32) {
   a[id] = 1;   //  If indices are clamped, then at least one race will occur. 
                     // 11 writes to 10 locations will cause at least one race.
}
```

The same pattern can occur for storage buffers.

The way the spec is written now, a careful WGSL programmer will anticipate this problem and predicate their writes, i.e.  write **their own** code to implement *just* the third rule.
That may seem too unreasonable for WGSL.

Second, if you think attempted OOB writes should be disabled for *shared* memories, then there's an argument that the same rule should apply across all address spaces.

Additional notes:
* if I recall correctly, D3D disables OOB writes to buffers
* Vulkan robustness2 disables OOB writes to buffers
* I don't have specific info on Metal

**Comment by dneto0** [comment]
Let's think about reads too.
Can you generate races, because the clamping messes with indices for writes or reads.  And accesses that would not ordinarily conflict may end up conflicting after you clamp indices.

**Comment by dneto0** [comment]
> Can you generate races, because the clamping messes with indices for writes or reads. And accesses that would not ordinarily conflict may end up conflicting after you clamp indices.

Yes. This can happen.

**Comment by teoxoy** [comment]
I think this should cover storage textures as well (since we recently decided to support read-write storage textures). 

Currently the spec allows the `textureLoad` fn to return "The data for some texel within bounds of the texture"; and
`textureStore` fn to "store value to some in bounds texel" which I think we can remove/disallow.

# Write

OOB writes to storage textures are always ignored on all target APIs.

References:

### Metal

> For all texture types, an out-of-bounds write to a texture is ignored.

_from page 184 of [Metal Shading Language Specification](https://developer.apple.com/metal/Metal-Shading-Language-Specification.pdf)_

### Vulkan

> The integer texel coordinates are validated according to the same rules as for texel input [coordinate validation](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap16.html#textures-integer-coordinate-validation).
> If the texel fails integer texel coordinate validation, then the write has no effect.

_from [16.4.2. Integer Texel Coordinate Validation](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap16.html#textures-output-coordinate-validation)_

### D3D

> Out of bounds addressing means nothing gets written to memory.

_from [direct3dhlsl docs](https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/store-uav-typed--sm5---asm-#remarks)_

# Read

OOB reads should also not read from the texture.

API specific behaviors + solutions (covering reads):

### Vulkan

We can rely on `robustImageAccess` [+ an additional check on the level param](https://github.com/KhronosGroup/Vulkan-Guide/blob/main/chapters/robustness.adoc#vk_ext_image_robustness) or `robustImageAccess2` if available or inject the check ourselves.

### D3D

We get this for free.

> When one or more of the coordinates in Location exceed the u, v, or w mipmap level dimensions of the texture, Load returns zero in all components. Direct3D guarantees to return zero for any resource that is accessed out of bounds.

_from [direct3dhlsl docs](https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-to-load)_

### Metal

This is odd, textures with color formats return well defined values but that's not the case for depth/stencil formats.
So, we unfortunately we need to inject the check ourselves since reads from storage textures with depth/stencil formats return a vector of the form `(0, U, U, U)` where `U` = undefined value.

![image](https://user-images.githubusercontent.com/28601907/222780730-1f73cd19-ab11-4d3c-870f-5a596195f7d9.png)

_from [Metal Shading Language Specification](https://developer.apple.com/metal/Metal-Shading-Language-Specification.pdf)_

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2023-03-07-minutes#wgsl-consider-disabling-out-of-bounds-memory-accesses-to-shared-memories-3893">WGSL 2023-03-07 Minutes</a></summary>

* BC:  A partner was converting their shaders.  One of their shaders was rejected by FXC. FXC said it was a data race.  We investigated, and spotted that bounds clamping could produce a data race on the write.  FXC in this case could see it deterministically.  HLSL disables writes when they are out of bounds.
* BC: Our software robustness was taking something that was sound in HLSL terms and turning it into a data race. It turns out the partner had messed up, subtracting a giant number from all IDs. FXC saw that all indices were zero, and balked.  They fixed their problem and the problem went away.
* BC: It was enough of a concern that we wrote the code to do predication instead. I implemented this, and then found a bunch of driver bugs from this.  Fortunately many of those driver bugs have been fixed with more recent drivers. But others are still open, and predication is broken.  It’s very concerning. We will continue to clamp for v1 launch.  It’s easier to go from clamping to predication than the other way around. If Chrome ships with predication, then people will rely on that (Hyrum’s law), and then switching to clamping would break shaders.  We’re more comfortable with clamping first, then switching to predication later.
* TR: Yeah, FXC would only have that error if it could tell it was constant, and knew that all threads were writing to the same value. Or a uniform value.  FXC would be ok with it if it saw that the index might not hit the bounds.
* AB: If you do clamping in software, we say “you get one of these X number of results”.  But really it’s undefined behaviour in a very common case. It’s a bit of an awkward truth. We’re currently doing investigation in Vulkan about what’s reliable in terms of their robustness feature. It’s a different thing between the underlying HW doing something vs. a tool generating the conflicting case. Will come back with the Vulkan answer.
* TR: The no-writing OOB behaviour is limited to things with descriptors. (e.g. buffers, but not with things bound through root arguments in the root signature)  There’s no bounds checking on (ByteAddressBuffer, and StructuredBuffer as root arguments)  [Root sig 1.1](https://learn.microsoft.com/en-us/windows/win32/direct3d12/root-signature-version-1-1) has an option where a driver is allowed to move something from descriptor to root signature, and then lose the bounds checking. \
See: [D3D12_DESCRIPTOR_RANGE_FLAGS (d3d12.h) - Win32 apps | Microsoft Learn](https://learn.microsoft.com/en-us/windows/win32/api/d3d12/ne-d3d12-d3d12_descriptor_range_flags) \
For: D3D12_DESCRIPTOR_RANGE_FLAG_DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS = 0x10000
* AB: Ok, as long as it’s only on deliberate request.
* RC: Echoing what Tex said.   Through descriptor, know the size and it gets bounds.  Through root signature, it’s just a pointer without a size.  A later version of root signature there is an additional option even more that says “if you do move that to root signature, do also retain the size and the bounds checks.”
* RC: There’s no bounds checks on other kinds of arrays in your shader, e.g. groupshared memory.
* BC: Mapping that to WGSL terms, that’s workgroup, private, function. 
* AB: Similar in Vulkan you only get robustness supports on buffer resources.
* MM: About the original topic.  WebGPU will have data races. Presumably whoever writes an HLSL implementation will need to have some mitigations. Is the issue that FXC detects the data race and there’s no way to stop it from doing so?
* BC: At the start the concern was getting a bunch of shader compilation failures. But now we know there was some developer error involved here, shader compilation failure is not our concern.
* BC: Spec now has several options on OOB writes and reads.  
* AB: Clarifications: We say anything that’s in the buffer that’s attached.  If it’s local to the GPU its any memory inside the variable’s space.  The UB from data races is not just about the data you get back, but also the assumptions the compiler makes about the kinds of values that are read back; will affect downstream code.
* MM: We’ve known about data races for a long time now, so why is this new?
* AB: My issue is the spec is misleading right now.  It’s still incumbent on the implementations to make sure the actions of the machine are still secure. That’s a separate thing.
* AB: Myles is correct a user can introduce their own data race, and that will be UB on the underlying platform. There’s only so many mitigations.  My suggestion is to be more honest with readers of the spec.  By weakening it, it’s a signal to be more cautious.
* **(Tabled)**


</details>

**Comment by dneto0** [comment]
> I think this should cover storage textures as well (since we recently decided to support read-write storage textures).

I think treatment of textures is interesting, but I'd like to split that to its own issue.

The discussion otherwise, and from the 2023-03-14 meeting has been about accesses via references and their associated pointers; not about via 'handles' for textures.

The resolution from the March 14 meeting is:
- out of bounds access is a **dynamic error**, for every address space.
    - in particular, it's considered an author's error, something they should not do.
- out of bounds is with respect to the particular array type.
- downgrade the existing list of limiting behaviours to an informative note to display a possible array of behaviours
- the list of behaviours should call out the data race case
- the list of behaviours should include trapping #3272
         * > Trapping is defined as causing the shader stage to immediately end, feeding zeroes to all pipeline stage output variables.
- create a new issue to define the consequences of a "dynamic error". #3954

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2023-03-14-minutes#wgsl-consider-disabling-out-of-bounds-memory-accesses-to-shared-memories-3893">WGSL 2023-03-14 Minutes</a></summary>

* KG: Sounded like the three things are:
    * Should probably resolve and look forward to editorial change to make spec scarier.
    * Question to AB if we should wait on more news from Vulkan or if we have all relevant news
* AB: Don't think it will be quick.
* KG: But that it would effect us?
* AB: Of course it might. I can't predict how the resolution will go.
* KG: Can pretend we think it will be fine
* DN: Looking at points from last discussion, thrust is some of these behaviours we say in the spec induce a data race and lets be honest and say it's a data race and will note that a user who has written code like this and will get wrong results and should say that and this also subsumes trapping behaviour as one of the UB behaviours is trapping. Noticed in the spec we say it's dynamic error if it's unchecked at runtime and spec says webgpu specs says what happens and it doen't . So this is an opportunity to fill that in. Opportunity to make internally consistent, call dynamic error and then WGSL spec doesn't say much about what program does, WebGPU spec may say more.
* KG: For the trapping behaviour, where we left off was MM was going to make a proposal
* MM: Yup, not done yet.
* KG: Doesn't sound like an easy proposal.
* MM: Quick note on what DN said, don't think material, but important, spec shouldn't say Undefined Behaviour. That's a trigger word, cal it something else and that's Ok.
* DN: I'ts a Dynamic Error if index  is outside bounds then find a place to say the shader may or not behave according to … might get device lost. Can't remember what it says you can touch. Don't need to put phrase undefined behaviour
* KG: Can always get Context lost.
* MM: Any situation where context gets lost is a situation which might violate the security model. Don't think we can get context lost. The context would be lost if you tried to preform the memory access out of bounds and if you did that there is a chance the security guarantee is violated so claiming any security mitigation all browsers will have to do will require as a side effect that the device won't be lost. Many other things could happen, but dont' think device lost could happen.
* KG: Don't think I agree, but disagree in a non-useful way. Think we're on the same page, but not sure if i agree with that
* MM: Maybe we intentionally choose to loose the device, but that's a different thing
* AB: One other point after looking at more stats on Spirv side. Disagreement in Vulkan, spirv- says scary UB to "form" a UB pointer. So any mitigation has to gate if any pointer is formed.
* MM: Just forming the pointer?
* AB: No agreement on if robustness fixes this.
* JB: Forming pointer restriction is same as in c++. In c++ if you form a pointer beyond bounds it's UB.
* KG: Can go 1 past and 1 before and even forming a .. pointer is ub. In practice mostly works out until compilers get smarter. The spir-v discussion will be along the lines of can we agree to do better then this and if the answer is no then we have to care
* AB: Someone might not be confident enough of all of their old architectures but that's a simple way to boil it down. As these things go across hardware generations it's hard to know how much of the market.
* KG: So, almost good news. Could have an extension saying new implementation has guarantee and anyone that does make that guarantee could use that extension. Could we plant the trees under which our children will shade themselves?
* AB: Discussion is progressing but tend to take a long time as it's a subtle discussion.
* JB: Question for AB on understanding SPIR-V spec, when it says UB in OpAccessChain spec, c++ has undefined behaviour, but POSIX says processes are contained and OS has security model to keep in process. So, c++ UB is bounded. Whats external bound of this UB behaviour
* AB: Vulkan could choose to constrain in the environment specification. One fo the two parts of question is do the robust buffer and image features fix the UB of the op access chain. That is not an immediate yes or not. 
* JB: So, if the Vulkan group doesn't know what the agreement was then we need to fall back to the OS vendors and say, does linux enforce this, does windows enforce between processes?
* AB: Don't know if the OS could do this because you're putting it over PCIe bus. Have own memory you've transferred in and page separation doesn't apply
* JB: So up to driver author?
* AB: Hardware architecture
* KG: Buck stops at IHV.
* JB: So up to authors of drivers to know what hardware will do.
* KG: Part of this is knowledge gathering and if we can figure out how to deal with it that's tolerable.
* MM: From first principles, all in agreement that clamping should be a viable thing that browser do to array and buffer indices and all in agreement that doing clamping can cause UB. That's a fact. So seems legit for the spec to say doing out of bounds might cause a data race.
* KG: Yup, think we're on board for that. Only other thing was someone mentioned saying it would be nice if we were able to say we didn't need clamping and could better define this behaviour entirely. But we should, regardless of appetite for that, don't remove clamping from v1.
* MM: Think that's right, the way you uphold without clamping is with branches and that has a different performance characteristic and don't think we should mandate that today. Maybe in the future, but right now it's the wrong solution
* KG: Resolution, make spec sufficiently scary and establish we're keeping clamping for v1 and waiting more news from Vulkan and Spir-v.
* DN: that's clear, are we going to discuss if this is a user problem, dynamic error?
* KG: The data race that caused it?
* DN: For the data races and for private/function scope.
* KG: For data race we have this language, just not clear that out of abounds can cause data race. So add the scary thing "if you access outside buffer can cause data races, data races are dynamic errors."
* DN: That's pretty good already
* MM: Think it's a bit past that, imagine you have buffer and inside is struct and inside is 2 arrays that have length. Should be legal to clamp to individual arrays instead of entire buffer. Out of bound does not mean out of bounds of buffer, but out of bounds of variable.
* KG: … type?
* DN: agreed
* MM: It's also possible to turn everything into points and the variable goes away, and so both options have to be valid.
* KG: The dynamic error is as soon as you access outside the logical bounds of the variable.
* DN: Should we still call it dynamic error for non-shared memory, function and private scope. Consistency argument, but data-race motivation doesn't hold.
* AB: If want to make dynamic error and programming error, not security issues and could remove clamping and make dynamic error and leave on user. Not portability argument, but don't have to inject code, maybe perf differences.
* MM: From security model, function scoped array, with dynamic index could leak into global memory and if don't clamp, have violated security boundary so think you need even for function scoped.
* DN: YOu do, but do we consider it a dynamic error. Can they just rely on the clamping or anything writes?
* JB: Your point is, right now it isn't an error to be out of bounds in those spaces and will we make it an error for consistency.
* MM: Feel like we should be as punitive as possible for v1, and can relax later.
* DN: I like it, that's the outcome I want. Think we should consider it a programmer error if you index out of bounds.
* AB: In my mind, remove all text and don't discuss behaviours and just say it's a dynamic error. there are multiple levels, what you read in spec is not what's being guaranteed.
* KG: Right, this is what gives you trapping, you just get it.
* MM: Historically this group has wanted a list of behaviours for the sake of authors, right?
* DN: Could be a note. You many get this or this or something else. It's a portability hazard.
* KG: Think those are helpful, but historically questions around testability why you want one of n behaviours and trapping wasn't one of them. Was imagining the minimum viable trapping was execution may stop. And that would be compatible with making it dynamic error. So, as long as no-one is opposed, sounds like everyone in the room is good? 
* DN: What's the that?
* KG:  **Indexing out of bounds is dynamic error**
* **JB: For all address spaces**
* MM: We have to have something that explains dynamic error, can't have term without text
* DN: Sure.
* KG: For out of abounds, non-normative note on you should expect read 0, or something else, or this other thing.
* MM: Not even that, can get values that don't' appear in memory, that's legal.
* KG: Sure.
* MM: So, when you do a read you get any value regardless of it's in memory. The only requirement is it can't come from outside the security boundary
* KG: We say you can get any value. I'd say it, it can be any number from anywhere, the security boundary is around this and overrides that to, I said anywhere but within the security boundary that applies.
* DN: Spec has indeterminate value, any bit pattern cast to type. It's garbage data. Doesn't matter security boundary as we don't define that thing. Have high level agreement on what happens. Split off what dynamic error is and maybe that's in WebGPU API. Maybe split in to 2 discussions
* AB: Feels like Indeterminate value isn't quite right in how we defined it. It's a frozen undefined value. A dynamic error has non-local effects. That's the key thing, you have non-local effects. WOuldn't call it non-determinate as we want that for other things.
* MM: Also if you say data races cause indeterminate values doesn't cover the case of program stops running.
* KG: Data races are not supposed to cause execution to cease. Enough for resolution?
* MM: Don't have feelling if it should be in this or that document but needs to be described somewhere
* DN: Will file spec to define dynamic error
* KG: Index out of bounds is dynamic error, DN will define dynamic error, dynamic error for all address spaces.


</details>
